### PR TITLE
TCP Fast Open support for ss-tunnel

### PR DESCRIPTION
**Pull request by PantherJohn**
`delayed_connect_cb` is not used here because it adds latency to normal connections. 

TODO:

- [x] macOS support.
- [x] Windows support.

### CODE REVIEW

**Comment by madeye** [comment]
Please follow local.c to fix issues on Windows and macOS.

**Comment by PantherJohn** [comment]
Sorry I was in a rush. Can you keep this pull request open so that I can add cross-platform support in the future? Kind of ignored Windows and MacOS.

File: src/tunnel.c [hunk 1]
```diff
@@ -469,9 +475,104 @@ remote_send_cb(EV_P_ ev_io *w, int revents)
         close_and_free_server(EV_A_ server);
         return;
     } else {
+
```
Comment (madeye) [inline src/tunnel.c]: trailing whitespace

File: README.md [hunk 2]
```diff
@@ -1,4 +1,5 @@
 # shadowsocks-libev
+with TCP Fast Open enabled
```
Comment (madeye) [inline README.md]: Please don't touch this file.

File: src/tunnel.c [hunk 3]
```diff
@@ -469,9 +475,104 @@ remote_send_cb(EV_P_ ev_io *w, int revents)
...
+                    close_and_free_remote(EV_A_ remote);
+                    close_and_free_server(EV_A_ server);
+                }
+                return;
+            }
+        } else {
+            s = send(remote->fd, remote->buf->data + remote->buf->idx,
                 remote->buf->len, 0);
+        }
+
```
Comment (madeye) [inline src/tunnel.c]: Ditto.

File: src/tunnel.c [hunk 4]
```diff
@@ -693,18 +796,23 @@ accept_cb(EV_P_ ev_io *w, int revents)
...
-        return;
+        if (r == -1 && errno != CONNECT_IN_PROGRESS) {
+            ERROR("connect");
+            close_and_free_remote(EV_A_ remote);
+            close_and_free_server(EV_A_ server);
+            return;
+        }
     }
-
+
```
Comment (madeye) [inline src/tunnel.c]: trailing whitespace

**Comment by PantherJohn** [comment]
> Please don't touch this file.


It's from my forked repo, so please ignore any changes in `README.md`

**Comment by PantherJohn** [comment]
added:

- [x] macOS support.
- [x] Windows support.

**Comment by PantherJohn** [comment]
@madeye any thoughts?

**Comment by madeye** [comment]
Thanks, merged.

**Pull request merged**